_posts/2025-06-04-golang-template-with-opentelemetry-support.md

```diff
@@ -0,0 +1,430 @@
+
+```sh
+"postgres://<user>:<password>@<host>:<port>/<dbName>?sslmode=<sslmode>"
+```
+
+Before you deploy the function to OpenFaaS make sure the secret exists. This can be done with the faas-cli:
+
+```sh
+faas-cli secret create pg-connection \
+  --from-file .secrets/pg-connection
+```
+
+Update the `stack.yaml` file:
+
+```diff
+functions:
+  employee-api:
+    lang: golang-otel
+    handler: ./employee-api
+    image: ttl.sh/employee-api:latest
++   secrets:
++     - pg-connection
+```
+
+### Add instrumentation
+
+To get traces for the database queries we will be using two instrumentation libraries:
+
+- [splunksql](https://github.com/signalfx/splunk-otel-go/tree/main/instrumentation/database/sql/splunksql) - Instrumentation for database/sql
+- [splunkpq](https://github.com/signalfx/splunk-otel-go/tree/main/instrumentation/github.com/lib/pq/splunkpq) - Instrumentation for github.com/lib/pq that uses splunksql
+
+> To find instrumentation libraries for the packages and frameworks used by your application you can search the [OpenTelemetry Registry](https://opentelemetry.io/ecosystem/registry/?language=go&component=instrumentation).
+
+Add the instrumentation packages to the function.
+
+```sh
+# Move into the handler directory 
+cd employee-api
+
+# Add required packages
+go get \
+  "github.com/signalfx/splunk-otel-go/instrumentation/database/sql/splunksql" \
+  "github.com/signalfx/splunk-otel-go/instrumentation/github.com/lib/pq/splunkpq"
+```
+
+Only a couple of small changes to the function are required to get traces for database queries using the instrumentation libraries we just added.
+
+Update the imports to add `splunksql` and swap out `lib/pq` for `splunkpq`:
+
+```diff
+package function
+
+import (
+	"database/sql"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net/http"
+	"os"
+
++	"github.com/signalfx/splunk-otel-go/instrumentation/database/sql/splunksql"
+
+-	_ "github.com/lib/pq" // PostgreSQL driver
++	_ "github.com/signalfx/splunk-otel-go/instrumentation/github.com/lib/pq/splunkpq"
+)
+
+```
+
+Update the `init` function and use `splunksql` to open the database connection.
+
+```diff
+func init() {
+	data, err := os.ReadFile("/var/openfaas/secrets/pg-connection")
+	if err != nil {
+		log.Fatalf("failed to read pg-connection secret: %v", err)
+	}
+	connStr := string(data)
+
+-	db, err = sql.Open("postgres", connStr)
++	db, err = splunksql.Open("postgres", connStr)
+	if err != nil {
+		log.Fatalf("Error opening database: %v", err)
+	}
+
+	if err := db.Ping(); err != nil {
+		log.Fatalf("Error connecting to the database: %v", err)
+	}
+
+	log.Println("Successfully connected to the database")
+}
+```
+
+This last step is important. Queries should be able to inherit the parent span and create a nested span to track work in a nested operation. In golang spans are passed using a `context.Context`. Make sure your code is passing on the context when making database queries, e.g. use `db.QueryContext`.
```

Comment:
Is is worth adding detail or linking out to something that explains spans prior to its use here?

Reply:
I added a short explanation of what spans are and their function. There already was a link to the OpenTelemetry docs on the bottom. The link has been moved up to the start of this section.